src/librbd/crypto/Types.cc

```diff
@@ -0,0 +1,46 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "Types.h"
+
+#include <string_view>
+#include "common/Formatter.h"
+
+namespace librbd {
+namespace crypto {
+
+void ParentCryptoParams::encode(bufferlist& bl) const {
+  ENCODE_START(1, 1, bl);
+  uint32_t key_size = wrapped_key.size();
+  encode(key_size, bl);
+  bl.append(wrapped_key.data(), key_size);
+  encode(block_size, bl);
+  encode(data_offset, bl);
```

Comment:
We don't open and parse parent header when loading.
With block_size and data_offset we have all the information we need.

Reply:
Was there a problem with opening the encryption header of the parent image?  Or just for convenience?

Having the same information stored on-disk twice is generally not desirable.

Reply:
Opening a header is costly:
first, you need to read the header from the OSDs
But even worst, you have to run pbkdf to parse the header, which costs a lot of CPU and time (takes order of seconds).
So if you have a long chain of formatted clones, opening the header of each ancestor on the way is a non-scalable solution.

Reply:
> But even worst, you have to run pbkdf to parse the header, which costs a lot of CPU and time (takes order of seconds).

I don't think the block size and data/payload offset information is obfuscated in any way.  For LUKS1 the location of the data/payload offset is fixed, so it is just a matter of reading 4 bytes.  For LUKS2 it is probably in JSON so it should be a matter of reading and parsing a bit of JSON, nothing beyond that.  Does libcryptsetup really not provide an API for getting these pieces of metadata without fully opening the LUKS container?

And what about the cipher?  Right now only two are supported and you choose between them based on the size of the wrapped key.  How would this work if support for a new cipher is added?

Reply:
> We don't open and parse parent header when loading.

So `.rbd_encryption_parent_cryptor` metadata acts as a make-do header and the real LUKS header is entirely side-stepped for parent images?  This means that we are missing out on the security features built into LUKS.  One of them is the inflation of the size of the encrypted key for anti-forensic purposes (AFsplit).  The other is secure checksumming of decrypted keys (LUKS absolutely needs it for passphrase validation but it is a nice-to-have even if you "know" the key because if something goes wrong we would not only serve garbage to the user on reads but also allow them to corrupt their data on writes).

Were any approaches that don't require storing sensitive data on the side considered?

> So if you have a long chain of formatted clones, opening the header of each ancestor on the way is a non-scalable solution.

A long chain of formatted clones is almost always a bad idea so I'm not sure if the term scale applies here.  I don't see a problem with opening a reasonable number of headers (IMO reasonable is 1-3).

Reply:
Right, so the first solution we considered was just to open the LUKS header per each ancestor.
There were a couple of drawbacks here:
1. Added cost for opening multiple headers (this is obviously a client requirement question)
2. Added complexity from client-side perspective, to remember a full chain of passphrases per image. Note that clients using clones usually don't need to care about the COW structure of their image, as RBD handles it transparently for them.
3. More complex API for loading encryption, e.g. supplying some sort of mapping between image_spec -> passphrase. This is not just for RBD itself, but for the entire ecosystem, e.g. qemu/libvirt.

So after thinking about these, we came up with the idea to wrap the parent key using the child key.
This gives a very neat and clean API for the user.
It does not require big changes to qemu and libvirt API, and it actually fits nice with libvirt clone API, as libvirt clone already takes in as arguments an encryption spec for both the source image and the dest image.

It's true that we're missing 2 luks features, AFSplit and secure checksumming.
However, I don't think these two are critical for us when we have RADOS as backend (LUKS was designed for a possibly failing HDD).
Anyway, from a client-side perspective, some clients will prefer this approach (while others may prefer the "conservative" approach).
The two approaches can live side-by-side in the future. For now, we decided to implement only this approach.

Reply:
> Right, so the first solution we considered was just to open the LUKS header per each ancestor. There were a couple of drawbacks here:
> 
> 1. Added cost for opening multiple headers (this is obviously a client requirement question)

Security is never free and performance is usually at the top of the trade off list.

> 2. Added complexity from client-side perspective, to remember a full chain of passphrases per image. Note that clients using clones usually don't need to care about the COW structure of their image, as RBD handles it transparently for them.
> 3. More complex API for loading encryption, e.g. supplying some sort of mapping between image_spec -> passphrase. This is not just for RBD itself, but for the entire ecosystem, e.g. qemu/libvirt.

I'm not suggesting that users supply the entire list of passphrases -- I think that a single passphrase for the image that is being opened by the user is a must-have.  But I also feel that the current `.rbd_encryption_parent_cryptor` metadata design, where it not only contains redundant information but also stores the volume key outside of the LUKS framework, is bad.
> 
> So after thinking about these, we came up with the idea to wrap the parent key using the child key. This gives a very neat and clean API for the user. It does not require big changes to qemu and libvirt API, and it actually fits nice with libvirt clone API, as libvirt clone already takes in as arguments an encryption spec for both the source image and the dest image.
> 
> It's true that we're missing 2 luks features, AFSplit and secure checksumming. However, I don't think these two are critical for us when we have RADOS as backend (LUKS was designed for a possibly failing HDD). Anyway, from a client-side perspective, some clients will prefer this approach (while others may prefer the "conservative" approach). The two approaches can live side-by-side in the future. For now, we decided to implement only this approach.

Well, RADOS is going to continue to run on HDDs for a very long time to come...

Are you aware of LUKS2 unbound keyslots?  IIUC an unbound keyslot allows storing an arbitrary key (i.e. one that is not used for encrypting any part of user data) with key wrapping, AFsplit, checksumming, etc all provided by the LUKS framework.  If we limit the format of formatted clones to LUKS2 (entirely reasonable IMO), I think we could store the parent volume key there and get rid of `.rbd_encryption_parent_cryptor`. I don't know whether libcryptsetup API is rich enough for this, but even if it currently isn't, it would be a much better solution than `.rbd_encryption_parent_cryptor` because the parent volume key in the child image would be stored exactly the same way and with the same level of security as in the parent image and nothing would be stored on the side.

Reply:
Storing the parent key in a keyslot was my first preference.
I tried concatinating the parent key to the child key, but libcryptsetup verifies the key length matches the cipher, and fails it.
I don't think you can store arbitratry data in keyslots, but rather keyslots are just same key encrypted using different means (e.g. different passphrase).
At the time when we considered alternatives, together with Jason about ~1 year ago, I digged the LUKS specs, as well as libcryptsetup API and found no other better alternative.

Of course feel free to double check me:
https://github.com/mbroz/cryptsetup/blob/master/lib/libcryptsetup.h

BTW, my experience with asking for API changes from libcryptsetup folks was not good.
e.g. https://gitlab.com/cryptsetup/cryptsetup/-/issues/600 (when we tried to avoid the memfd_create "hack" we currently have in the  librbd code)

Reply:
> Storing the parent key in a keyslot was my first preference.
> I tried concatinating the parent key to the child key, but libcryptsetup verifies the key length matches the cipher, and fails it.

I'm going by the on-disk format spec (https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/docs/on-disk-format-luks2.pdf):
```
Keyslot can contain an unbound key (key not assigned to any encrypted data segment)
that can be used for external applications. Size of an unbound key can be different from
volume key size in other keyslots.
```
Note that this functionality is LUKS2-only.  Could it be that you were trying that with LUKS1?

Reply:
> BTW, my experience with asking for API changes from libcryptsetup folks was not good. e.g. https://gitlab.com/cryptsetup/cryptsetup/-/issues/600 (when we tried to avoid the memfd_create "hack" we currently have in the librbd code)

I'm hoping that this case is different since the spec explicitly mentions external applications and skimming though libcryptsetup.h I see doc comments suggesting that at least some thought was put into exposing the unbound keyslot functionality in the API.

Reply:
The unbound key option generates its own key. It does not let you store your own key.
Given this, do you still see a way to use this feature?

Reply:
> The unbound key option generates its own key. It does not let you store your own key.

I haven't actually tried it but this doesn't match my reading of the code:
```
int crypt_keyslot_add_by_key(struct crypt_device *cd,
        int keyslot,
        const char *volume_key,
        size_t volume_key_size,
        const char *passphrase,
        size_t passphrase_size,
        uint32_t flags)
{
        ...
        if (volume_key)
                vk = crypt_alloc_volume_key(volume_key_size, volume_key);
        else if (flags & CRYPT_VOLUME_KEY_NO_SEGMENT)
                vk = crypt_generate_volume_key(cd, volume_key_size);
        else if (cd->volume_key)
                vk = crypt_alloc_volume_key(cd->volume_key->keylength, cd->volume_key->key);
        else
                return -EINVAL;

        if (!vk)
                return -ENOMEM;
```
If `volume_key` is supplied, it seems to use it -- `crypt_generate_volume_key` isn't called.  Or am I looking at the wrong function?

> Given this, do you still see a way to use this feature?

Yes, like I said, I would try to go as far as attempting to patch libcryptsetup.

Reply:
(The comment I was replying to disappeared.)

Reply:
> > Storing the parent key in a keyslot was my first preference.
> > I tried concatinating the parent key to the child key, but libcryptsetup verifies the key length matches the cipher, and fails it.
> 
> I'm going by the on-disk format spec (https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/docs/on-disk-format-luks2.pdf):
> 
> ```
> Keyslot can contain an unbound key (key not assigned to any encrypted data segment)
> that can be used for external applications. Size of an unbound key can be different from
> volume key size in other keyslots.
> ```
> 
> Note that this functionality is LUKS2-only. Could it be that you were trying that with LUKS1?

Indeed I missed that option.

So how about this:
Instead of storing ParentCryptoParams as RBD metadata, we will store it (minus the AES-wrap which will no longer be needed) as an unbound keyslot.
Actually, we will store possibly multiple ParentCryptoParams, one for each formatted ancestor.
This is to keep a cost of a single pbkdf per image.encryption_load.
We will also put the child key there (in addition to the "bound" keyslot, which will exist since libcryptsetup.format does not allow to skip it, but will not be read by us).

Reply:
> Instead of storing ParentCryptoParams as RBD metadata, we will store it (minus the AES-wrap which will no longer be needed) as an unbound keyslot.
> Actually, we will store possibly multiple ParentCryptoParams, one for each formatted ancestor.
> This is to keep a cost of a single pbkdf per image.encryption_load.

My first concern is that you would be taking something inherently recursive (a chain of parents where each parent image stores its own metadata on its own) and linearizing (part of) that into each individual clone.  I don't think librbd sets a limit on the length of the parent chain and you would likely need one because the space in the keyslot isn't infinite.  But even with the limit in place, from a high-level perspective breaking recursiveness feels wrong.

The second larger concern is that you would be re-encrypting and storing the same volume key an unbounded number of times.  E.g. if you have 10000 VMs based off of the same golden image, its volume key would be stored in 10000 different places.  Each copy would be encrypted differently and given that the volume key itself is just 32 bytes long, I'm worried that this might be compromising its security by handing the attacker multiple different ciphertexts of the same plaintext.  This is true for the current `.rbd_encryption_parent_cryptor` metadata approach as well and this is the primary reason I picked on it.  But then I'm not a cryptography expert so I would be happy to be told otherwise by someone who knows better.

> We will also put the child key there.

I'm not sure what you mean here.  The volume key for the child image would be placed in a normal (i.e. bound) keyslot automatically, wouldn't it?

Reply:
> > Instead of storing ParentCryptoParams as RBD metadata, we will store it (minus the AES-wrap which will no longer be needed) as an unbound keyslot.
> > Actually, we will store possibly multiple ParentCryptoParams, one for each formatted ancestor.
> > This is to keep a cost of a single pbkdf per image.encryption_load.
> 
> My first concern is that you would be taking something inherently recursive (a chain of parents where each parent image stores its own metadata on its own) and linearizing (part of) that into each individual clone. I don't think librbd sets a limit on the length of the parent chain and you would likely need one because the space in the keyslot isn't infinite. But even with the limit in place, from a high-level perspective breaking recursiveness feels wrong.
> 
> The second larger concern is that you would be re-encrypting and storing the same volume key an unbounded number of times. E.g. if you have 10000 VMs based off of the same golden image, its volume key would be stored in 10000 different places. Each copy would be encrypted differently and given that the volume key itself is just 32 bytes long, I'm worried that this might be compromising its security by handing the attacker multiple different ciphertexts of the same plaintext. This is true for the current `.rbd_encryption_parent_cryptor` metadata approach as well and this is the primary reason I picked on it. But then I'm not a cryptography expert so I would be happy to be told otherwise by someone who knows better.

I'm trying to find proof for this.
Google found that people asked the same question:
https://crypto.stackexchange.com/questions/57720/can-sending-the-same-plaintext-twice-each-encrypted-with-different-key-and-iv
https://crypto.stackexchange.com/questions/9723/vulnerabilities-if-encrypting-the-same-data-with-2-different-keys

But I don't know how expert are the people that gave the answer. I will continue searching...

But anyhow, as you said before "single passphrase is a must". Do you see any other solution with a single passphrase that does not re-encrypt the parent secret?

BTW in LUKS itself, you can create upto 8/32 keyslots, containing the same key encrypted with different passphrases.

> 
> > We will also put the child key there.
> 
> I'm not sure what you mean here. The volume key for the child image would be placed in a normal (i.e. bound) keyslot automatically, wouldn't it?

Opening each keyslot incurs a pbkdf cost. I'm trying to minimize that cost.

Reply:
> The second larger concern is that you would be re-encrypting and storing the same volume key an unbounded number of times. E.g. if you have 10000 VMs based off of the same golden image, its volume key would be stored in 10000 different places. Each copy would be encrypted differently and given that the volume key itself is just 32 bytes long, I'm worried that this might be compromising its security by handing the attacker multiple different ciphertexts of the same plaintext. This is true for the current .rbd_encryption_parent_cryptor metadata approach as well and this is the primary reason I picked on it. But then I'm not a cryptography expert so I would be happy to be told otherwise by someone who knows better.

Regarding the security concerns. 
I first want to point out that the proposed method does not diminish security as opposed to the current best practice (and only option) – namely, to do a full copy of the data into a fresh volume (with a fresh key) for each cloned image. In the full copy scenario you have the exact same data encrypted potentially thousands of times, each time with a new key & passphrase pair. In our setting it is the same volume key or volume passphrase which is encrypted potentially thousands of times, but the security is virtually the same – break one and you get all of the parent image data. 

That being said, I want to address the concerns concretely (everything below is relevant to both full-copy or the proposed cloning method).   
There are two different aspects here:
1)	**Having a large number of encrypted versions of the same data, each with a different key** -  In terms of pure cryptographic security, this has a very limited impact. Similarly, if you encrypt the same block over and over again within the same data stream, it is not going to put this data at risk and is not forbidden by ciphertext designers. Generally speaking, attacks on AES do not know how to  benefit from known plaintext, or correlated plaintexts. In fact AES is assumed to be CPA-secure (secure against Chosen Plaintext Attacks) by which an attacker can ask for multiple encryptions of any plaintext of his choice, and still would not be able to distinguish if a given ciphertext is an encryption of a known plaintext or a random string (this is even when using the same key, let alone with different keys).  It may be possible in some cases to slightly reduce the time to do a brute-force attack by enumerating all potential keys and hoping to hit one of the potential cipher targets rather than a single target (this mainly applies to AES-CTR mode, and I don’t see how it helps with AES-XTS). In any case, such a reduction is insignificant in the grand scheme of things and will still leave us with a brute-force cost of over 2^200. 

2)	**Having multiple passphrases to open the same data volume** - The problem here is that the strength of the encryption is reduced to that of the *weakest* passphrase. With human chosen passphrases, this may be a concern. In a decent implementation (such as a cloud setting) the passphrases are randomly generated and stored in an HSM and are equally strong so this would not be an issue. In fact, LUKS supports such a multi-passphrase design by which the same volume key can be encrypted in multiple key-slots (albeit the limit there is 8 slots in LUKS1 and 32 slots for LUKS2). There too, the strength is reduced to that of the weakest passphrase and it boils down to using good practices with regards to passphrases.

Reply:
> In the full copy scenario you have the exact same data encrypted potentially thousands of times, each time with a new key & passphrase pair. In our setting it is the same volume key or volume passphrase which is encrypted potentially thousands of times, but the security is virtually the same – break one and you get all of the parent image data.

I'm not sure if drawing parallels between encrypting bulk data and key material is fair, both in general and here in particular.  Given a bunch of clones, an attacker wouldn't necessarily know whether they are looking at the same data or not.  At a given point in time, some clones would have overwritten 1% of the parent data, others - 50%.  And either way it would mean at least megabytes of data to sift through, potentially trying to identify some well known and more or less immutable blobs such as files in `/etc`, matching on-disk locations, etc.

A volume key OTOH would stay the same even after all of the parent data is overwritten, and provide access to parent data that is no longer visible otherwise.  And it is just 32 or 64 bytes, stored in a fixed location.

Do you know of any other storage systems that implement encrypted clones?  I know vSphere doesn't -- choosing a different encryption key for the clone isn't supported.

> There are two different aspects here:
> 1. **Having a large number of encrypted versions of the same data, each with a different key** [...]
> 2. **Having multiple passphrases to open the same data volume** [...]

These points are correct when considered on their own (although CTR and XTS don't seem to be relevant for Key Wrap), but the devil is always in the details.  For example, the proposed cloning scheme uses the volume key (DEK) of the child image as a wrapping key for the volume key of the parent image.  A given volume key is thus used for both data encryption and key wrapping.  Doesn't that violate a well known principle of key separation for symmetric keys ("one symmetric key should be dedicated to only one usage")?

Or the fact that here the parent volume key is wrapped and stored in the child image metadata as is, while libcryptsetup, before "wrapping" with the passphrase and storing in the LUKS header, would run it through AFsplit.

And of course the fact that we need to design for human chosen passphrases because few users would have the luxury to use an HSM or similar.

Reply:
Could you please put together a doc (e.g. `doc/dev/rbd-encryption.rst`) with a high-level problem description and a detailed explanation of what is stored where, how big are the keys, how the keys are wrapped, how the data is encrypted and re-encrypted on copyup, etc?  Basically something that folks who aren't going to look at the code (or even really familiar with RBD) could review and comment on the security aspects.